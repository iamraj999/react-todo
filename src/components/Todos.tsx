import * as React from "react";
import {
  AddTodo,
  Delete_Todo,
  Complete_Todo,
  loadTodos,
} from "../store/actionCreator";
import { connect } from "react-redux";
import { CircularProgress } from "@material-ui/core";
import { useState } from "react";

export const Todos = ({
  todos,
  loading,
  error,
  loadTodos,
  addTodo,
  deleteTodo,
  completeTodo,
}) => {
  const [title, setTitle] = useState("");
  function handleChange(e) {
    setTitle(e.target.value);
  }
  return (
    <main>
      <input
        id="title"
        placeholder="Add Todo title"
        value={title}
        onChange={(e) => handleChange(e)}
      ></input>
      <button
        className="addButton"
        onClick={() =>{ addTodo({ title })
      setTitle("")}}
        disabled={title ? false : true}
      >
        Add Todo
      </button>
      <button onClick={() => loadTodos()}>Load from api</button>
      {loading ? (
        <main>
          <CircularProgress />
        </main>
      ) : null}
      <ul className="todo-list">
        {todos.map((todo:any) => {
          const id = todo.id.toString();
          return (
            <div className="todo-container" key={id}>
              <input
                className="todo-checkbox"
                type="checkbox"
                onChange={(e) => {
                  todo.completed = e.target.checked;
                  completeTodo(e, todo);
                }}
                checked={todo.completed}
              />
              <button
                className="delete-todo"
                onClick={() => {const id = todo.id;
                  deleteTodo({id})}}
              >
                Delete
              </button>
              <li className={todo.completed ? "todo-completed" : ""}>
                {todo.title}
              </li>
            </div>
          );
        })}
      </ul>
    </main>
  );
};

const mapStateToProps = (state) => {
  return { todos: state.todos, loading: state.loading, error:state.error };
};
const mapDispatchToProps = (dispatch) => {
  return {
    addTodo: (todo) => dispatch(AddTodo(todo)),
    deleteTodo: (id) => dispatch(Delete_Todo(id)),
    completeTodo: (todo) => dispatch(Complete_Todo(todo)),
    loadTodos: () => dispatch(loadTodos()),
  };
};
export default connect(mapStateToProps, mapDispatchToProps)(Todos);
