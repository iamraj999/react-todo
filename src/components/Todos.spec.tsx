import { Todos } from "./Todos";
import * as React from "react";
import {
  AddTodo,
  Delete_Todo,
  Complete_Todo,
  loadTodos,
} from "../store/actionCreator";
import * as renderer from "react-test-renderer";
import { shallow } from "enzyme";

describe("Todos Component", () => {
  it("renders correctly", () => {
    const props = {
      todos: [],
      loading: false,
      error: "",
      loadTodos: jest.fn(),
      addTodo: jest.fn(),
      deleteTodo: jest.fn(),
      completeTodo: jest.fn(),
    };
    const wrapper = shallow(<Todos {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
  it("should disable the add todo button when no title is provided", () => {
    const props = {
      todos: [],
      loading: false,
      error: "",
      loadTodos: jest.fn(),
      addTodo: jest.fn(),
      deleteTodo: jest.fn(),
      completeTodo: jest.fn(),
    };
    const wrapper = shallow(<Todos {...props} />);
    expect(wrapper.find("button.addButton").props().disabled).toBe(true);
  });
  it("should enable the add todo button when title is provided", () => {
    const props = {
      todos: [],
      loading: false,
      error: "",
      loadTodos: jest.fn(),
      addTodo: jest.fn(),
      deleteTodo: jest.fn(),
      completeTodo: jest.fn(),
    };
    const wrapper = shallow(<Todos {...props} />);
    const event = {
      target: { value: "test" },
    };
    wrapper.find("input").simulate("change", event);
    expect(wrapper.find("button.addButton").props().disabled).toBe(false);
  });
});
