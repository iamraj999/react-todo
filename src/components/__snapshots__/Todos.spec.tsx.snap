// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todos Component renders correctly 1`] = `
<main>
  <input
    id="title"
    onChange={[Function]}
    placeholder="Add Todo title"
    value=""
  />
  <button
    className="addButton"
    disabled={true}
    onClick={[Function]}
  >
    Add Todo
  </button>
  <button
    onClick={[Function]}
  >
    Load from api
  </button>
  <ul
    className="todo-list"
  />
</main>
`;
